File: YourMainActivity.dart
import 'dart:convert';
import 'dart:io';
import 'package:flutter/material.dart';
import 'package:path_provider/path_provider.dart';
import 'package:http/http.dart' as http;

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: YourMainActivity(),
    );
  }
}

class YourMainActivity extends StatefulWidget {
  @override
  _YourMainActivityState createState() => _YourMainActivityState();
}

class _YourMainActivityState extends State<YourMainActivity> {
  Map<String, dynamic> cache = {};
  String cpuCode = "";
  String chipset = "";
  String gpu = "";

  @override
  void initState() {
    super.initState();
    loadCache().then((_) {
      cpuCode = getCPUInfo();
      getChipsetData(cpuCode);
    });
  }

  Future<File> _getCacheFile() async {
    final dir = await getApplicationDocumentsDirectory();
    return File('${dir.path}/chipset_cache.json');
  }

  Future<void> loadCache() async {
    try {
      final file = await _getCacheFile();
      if (await file.exists()) {
        cache = jsonDecode(await file.readAsString());
      }
    } catch (_) {
      cache = {};
    }
  }

  Future<void> saveCache() async {
    final file = await _getCacheFile();
    await file.writeAsString(jsonEncode(cache));
  }

  String getCPUInfo() {
    String raw = "UNKNOWN";
    try {
      final file = File("/proc/cpuinfo");
      if (file.existsSync()) {
        final lines = file.readAsLinesSync();
        for (var line in lines) {
          final lower = line.toLowerCase();
          if (lower.contains("hardware") || lower.contains("model name")) {
            raw = line.split(":")[1].trim();
            break;
          }
        }
      }
    } catch (_) {}

    return raw.toUpperCase().replaceAll(RegExp(r'[^A-Z0-9]'), '');
  }

  Future<void> getChipsetData(String cpuCode) async {
    if (cache.containsKey(cpuCode)) {
      setState(() {
        chipset = cache[cpuCode]["chipset"];
        gpu = cache[cpuCode]["gpu"];
      });
      return;
    }

    String url = "";
    if (cpuCode.startsWith("MT")) {
      url = "https://tolepcoy.pages.dev/tolepcoy_mediatek?code=$cpuCode";
    } else if (cpuCode.startsWith("MSM") || cpuCode.startsWith("SM")) {
      url = "https://tolepcoy.pages.dev/tolepcoy_qualcomm?code=$cpuCode";
    } else {
      setState(() {
        chipset = "Chipset not found!";
        gpu = "Unknown";
      });
      return;
    }

    try {
      final response = await http.get(Uri.parse(url));
      final data = jsonDecode(response.body);

      String fetchedChipset = data["chipset"] ?? "Nothing";
      String fetchedGpu = data["gpu"] ?? "Nothing";

      cache[cpuCode] = {"chipset": fetchedChipset, "gpu": fetchedGpu};
      await saveCache();

      setState(() {
        chipset = fetchedChipset;
        gpu = fetchedGpu;
      });
    } catch (_) {
      setState(() {
        chipset = "Chipset not found!";
        gpu = "Unknown";
      });
    }
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: Text("Chipset Info")),
      body: Padding(
        padding: EdgeInsets.all(16),
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.start,
          children: [
            Text("CPU Code: $cpuCode"),
            SizedBox(height: 8),
            Text("Chipset: $chipset"),
            SizedBox(height: 8),
            Text("GPU: $gpu"),
          ],
        ),
      ),
    );
  }
}

File: YourMainActivity.py
import os
import json
import re
import requests

CACHE_FILE = "chipset_cache.json"

if os.path.exists(CACHE_FILE):
    with open(CACHE_FILE, "r") as f:
        cache = json.load(f)
else:
    cache = {}

def get_cpu_info() -> str:
    cpu_info = ""
    try:
        with open("/proc/cpuinfo", "r") as f:
            for line in f:
                lower = line.lower()
                if "hardware" in lower or "model name" in lower:
                    cpu_info = line.split(":")[1].strip()
                    break
    except FileNotFoundError:
        cpu_info = os.environ.get("PROCESSOR_IDENTIFIER", "UNKNOWN")

    cpu_code = re.sub(r"[^A-Z0-9]", "", cpu_info.upper())
    return cpu_code or "UNKNOWN"

def get_chipset_data(cpu_code: str):
    if cpu_code in cache:
        return cache[cpu_code]

    if cpu_code.startswith("MT"):
        url = f"https://tolepcoy.pages.dev/tolepcoy_mediatek?code={cpu_code}"
    elif cpu_code.startswith("MSM") or cpu_code.startswith("SM"):
        url = f"https://tolepcoy.pages.dev/tolepcoy_qualcomm?code={cpu_code}"
    else:
        return {"chipset": "Chipset not found!", "gpu": "Unknown"}

    try:
        resp = requests.get(url)
        resp.raise_for_status()
        data = resp.json()

        chipset = data.get("chipset", "Nothing")
        gpu = data.get("gpu", "Nothing")

        cache[cpu_code] = {"chipset": chipset, "gpu": gpu}
        with open(CACHE_FILE, "w") as f:
            json.dump(cache, f, indent=2)

        return cache[cpu_code]

    except Exception:
        return {"chipset": "Chipset not found!", "gpu": "Unknown"}

if __name__ == "__main__":
    cpu_code = get_cpu_info()
    data = get_chipset_data(cpu_code)
    print(f"CPU Code: {cpu_code}")
    print(f"Chipset: {data['chipset']}")
    print(f"GPU: {data['gpu']}")